app/src/services/storage.js

```diff
@@ -35,37 +34,94 @@ const secretAccessKey = config.get('objectStorage.secretAccessKey');
+        if (bucketData.region) data.region = bucketData.region;
+      } catch (err) {
+        log.warn(err.message, { function: '_getBucket'});
+      }
+    }
+
+    return data;
+  },
+
+  /**
+   * @private
+   * @function _getS3Client
    * The AWS S3Client used for interacting with S3 compatible storage
+   * @param {string} options.accessKeyId The S3 Bucket accessKeyId
+   * @param {string} options.endpoint The S3 Bucket endpoint
+   * @param {string} options.region The S3 Bucket region
+   * @param {string} options.secretAccessKey The S3 Bucket secretAccessKey
    * @param {Readable} stream A readable stream object
    * @returns {object} A pre-configured S3 Client object
    */
-  _s3Client: new S3Client({
-    apiVersion: '2006-03-01',
-    credentials: {
-      accessKeyId: accessKeyId,
-      secretAccessKey: secretAccessKey,
-    },
-    endpoint: endpoint,
-    forcePathStyle: true,
-    region: 'us-east-1' // Need to specify valid AWS region or it'll explode ('us-east-1' is default, 'ca-central-1' for Canada)
-  }),
+  _getS3Client: ({ accessKeyId, endpoint, region, secretAccessKey } = {}) => {
+    if (!accessKeyId || !endpoint || !region || !secretAccessKey) {
+      log.error('Unable to generate S3Client due to missing arguments', { function: '_getS3Client'});
+    }
+
+    return new S3Client({
+      apiVersion: '2006-03-01',
+      credentials: {
+        accessKeyId: accessKeyId,
+        secretAccessKey: secretAccessKey
+      },
+      endpoint: endpoint,
+      forcePathStyle: true,
+      region: region
+    });
+  },
 
   /**
    * @function copyObject
-   * Creates a copy of the object at `copySource`
+   * Creates a copy of the object at `copySource` for the same bucket
    * @param {string} options.copySource Specifies the source object for the copy operation, excluding the bucket name
    * @param {string} options.filePath The filePath of the object
-   * @param {string} [options.metadata] Optional metadata to store with the object
-   * @param {string} [options.tags] Optional tags to store with the object
-   * @param {string} [options.metadataDirective=COPY] Optional operation directive
-   * @param {string} [options.versionId=undefined] Optional versionId to copy from
+   * @param {object} [options.metadata] Optional metadata to store with the object
+   * @param {object} [options.tags] Optional tags to store with the object
+   * @param {string} [options.metadataDirective=COPY] Optional metadata operation directive
+   * @param {string} [options.taggingDirective=COPY] Optional tagging operation directive
+   * @param {string} [options.versionId] Optional versionId to copy from
+   * @param {string} [options.bucketId] Optional bucketId
    * @returns {Promise<object>} The response of the delete object operation
    */
-  copyObject({ copySource, filePath, metadata, tags, metadataDirective = MetadataDirective.COPY, taggingDirective = TaggingDirective.COPY, versionId = undefined }) {
```

Comment:
i guess when we call copyObject from the metadata related controllers in controllers/object.js , we can also pass the bucket ID. and as you say, this function just copies within the same bucket.

Reply:
The `CopyObjectCommand` directive only works for same-bucket operations. While we do need to specify the bucketId so that we know which bucket to do the operation on, we won't be able to do cross-bucket copy operations as it is here. We'll have to figure out whether we want to make another service endpoint to explicitly do cross-bucket copies or if we want to extend this existing function to dynamically determine whether this is same-bucket or cross-bucket operation. TBD.